### PR TITLE
CompEngine features

### PR DESCRIPTION
**Pull request by FinYang**
FC_LocalSimple_lfit_taures, DN_OutlierInclude_abs_001_mdrmd need update. Waiting for response from Ben.

Using `acf`  instead of Fourier method in Matlab to calculate autocorrelation.

### CODE REVIEW

File: NAMESPACE [hunk 1]
```diff
@@ -1,12 +1,32 @@
...
+export(SY_SpreadRandomLocal_100_meantaul)
+export(SY_StdNthDer_1)
 export(acf_features)
 export(arch_stat)
+export(compengine)
 export(crossing_points)
 export(embedding)
 export(entropy)
 export(flat_spots)
+export(hctsa_autocorr)
```
Comment (robjhyndman) [inline NAMESPACE]: We don't need to preface the function names with `hctsa`. Just including a reference in the help file is enough.

File: R/compengine.R [hunk 2]
```diff
@@ -0,0 +1,780 @@
...
+#' @seealso \code{\link{AC_9}}
+#' @seealso \code{\link{CO_FirstMin_ac}}
+#' @seealso \code{\link{CO_trev_1_num}}
+#' @seealso \code{\link{SB_MotifTwo_mean_hhh}}
+#' @seealso \code{\link{PH_Walker_prop_01_sw_propcross}}
+#' @references B.D. Fulcher and N.S. Jones. hctsa: A computational framework for automated time-series phenotyping using massive feature extraction. Cell Systems 5, 527 (2017).
+#' @references B.D. Fulcher, M.A. Little, N.S. Jones Highly comparative time-series analysis: the empirical structure of time series and their methods. J. Roy. Soc. Interface 10, 83 (2013).
+#' @author Yangzhuoran Yang
+#' @export
+hctsa_autocorr <- function(x){
```
Comment (robjhyndman) [inline R/compengine.R]: This involves calling acf() many times which is not efficient. Also, we already have our own acf features function. Perhaps these can be combined?

**Comment by FinYang** [comment]
Ben mentioned C coding of the current features. Do you want an R wrapper for them or this is enough for now?

**Comment by robjhyndman** [comment]
I think we should first get the R code of the features done and working efficiently. Then we can think about whether any features would be better recoded in C.

**Comment by FinYang** [comment]
They already have the current C files at hand. I wouldn't ask for their C files then. 
There have been issues with the web domain so I haven't been able to upload the M4 data. Waiting for it to be fixed.

**Comment by FinYang** [comment]
The existing acf feature function only computes up to lag 10. This set of features requires much more. I've tried to avoid some acf calls.
Is there any efficient function that computes autocorrelation only at 1 lag at a time, like autocorr in MatLab,  instead of computing autocorrelation up to a max lag like `acf`? The order of autocorrelation that some features require depends on the computed result, so I thought a loop that can stop at some point might be more effecient than computing acf at all lags, especially when the series is big, unless it's even slower when it involves loops. The  `acf` is pretty fast anyway.

File: R/compengine.R [hunk 3]
```diff
@@ -0,0 +1,815 @@
...
+#' @seealso \code{\link{AC_9}}
+#' @seealso \code{\link{FirstMin_ac}}
+#' @seealso \code{\link{trev_num}}
+#' @seealso \code{\link{MotifTwo_entro3}}
+#' @seealso \code{\link{Walker_propcross}}
+#' @references B.D. Fulcher and N.S. Jones. hctsa: A computational framework for automated time-series phenotyping using massive feature extraction. Cell Systems 5, 527 (2017).
+#' @references B.D. Fulcher, M.A. Little, N.S. Jones Highly comparative time-series analysis: the empirical structure of time series and their methods. J. Roy. Soc. Interface 10, 83 (2013).
+#' @author Yangzhuoran Yang
+#' @export
+autocorr_features <- function(x){
```
Comment (robjhyndman) [inline R/compengine.R]: This is inefficient. Many of these functions call the acf() function on the same data.

File: R/compengine.R [hunk 4]
```diff
@@ -0,0 +1,815 @@
...
+          A[m] <- A[m] + 1
+          if (j < N){
+            B[m] <- B[m] + 1
+          }
+        }
+      } else{
+        run[jj] <- 0
+      }
+    }
+    for( j in 1:N-i){
```
Comment (robjhyndman) [inline R/compengine.R]: Be careful. Do you mean 1:(N-i) or (1:N)-i. I think this will give the latter.

File: R/compengine.R [hunk 5]
```diff
@@ -0,0 +1,815 @@
...
+  for(i in 1:(N-1)){ # go through each point in the time series, counting matches
+    y1 <- y[i]
+    for(jj in 1:(N-i)){ # compare to points through the rest of the time series
+      # Compare to future index, j:
+      j <- i + jj
+      # This future point, j, matches the time-series value at i:
+      if (abs(y[j]-y1) < r){
+        run[jj] <- lastrun[jj] + 1 # increase run count for this lag
+        M1 <- min(M, run[jj])
+        for (m in 1:M1){
```
Comment (robjhyndman) [inline R/compengine.R]: Loop not needed. Vectorize.

File: R/compengine.R [hunk 6]
```diff
@@ -0,0 +1,815 @@
...
+#' @return Standard deviation of the first derivative of the time series.
+#' @references cf. http://www.mathworks.de/matlabcentral/newsreader/view_thread/136539
+#' @references B.D. Fulcher and N.S. Jones. hctsa: A computational framework for automated time-series phenotyping using massive feature extraction. Cell Systems 5, 527 (2017).
+#' @references B.D. Fulcher, M.A. Little, N.S. Jones Highly comparative time-series analysis: the empirical structure of time series and their methods. J. Roy. Soc. Interface 10, 83 (2013).
+#' @author Yangzhuoran Yang
+#' @export
+Std1stDer <- function(y){
+  if(length(y)<2) stop("Time series is too short to compute differences")
+  yd <- diff(y)
+  return(sd(yd))
```
Comment (robjhyndman) [inline R/compengine.R]: What about missing values?

Comment (FinYang) [reply R/compengine.R]: I add a line next to it calculating the acf once
 `acfv <- stats::acf(x, length(x)-1, plot = FALSE)`
and then refer to the stored value using an argument whenever an acf call is required
`embed2_incircle_2 = embed2_incircle(x,2, acfv = acfv),`
so it won't repeatedly call the acf() function in each feature function.
Is this good enough? If not, could you give me some ideas?

File: R/compengine.R (hunk 3 above)
Comment (FinYang) [reply R/compengine.R]: I add a line next to it calculating the acf once
 `acfv <- stats::acf(x, length(x)-1, plot = FALSE)`
and then refer to the stored value using an argument whenever an acf call is required
`embed2_incircle_2 = embed2_incircle(x,2, acfv = acfv),`
so it won't repeatedly call the acf() function in each feature function.
Is this good enough? If not, could you give me some ideas?

Comment (robjhyndman) [reply R/compengine.R]: Yes, that will work.

**Comment by FinYang** [comment]
I'm not quite sure how to handle missing values since I don't understand some of the features completely. 
In `acf()`, na action is na.pass. In any simple arithmetic calculation, such as `mean`, `sd` and `sum`, missing values are removed. In function `cumsum` from function `fluctanal_prop_r1` where it implements fluctuation analysis (I don't know what that means), missing values are treated as 0.

**Comment by robjhyndman** [comment]
> I'm not quite sure how to handle missing values since I don't understand some of the features completely.
> In `acf()`, na action is na.pass. In any simple arithmetic calculation, such as `mean`, `sd` and `sum`, missing values are removed. In function `cumsum` from function `fluctanal_prop_r1` where it implements fluctuation analysis (I don't know what that means), missing values are treated as 0.

In `mean`, `sd` and `sum`, use `na.rm=TRUE`. `acf` is problematic as the most obvious option (using complete pairs) is not available. For now, perhaps use `na.action=na.contiguous`.

Anything using `cumsum`, treating missing as 0s is probably ok for now.

**Comment by robjhyndman** [comment]
> 3 points I've noticed that might worth mentioning
> 
> 1. When calculating points inside a given circular boundary in a 2-d embedding space in function `embed2_incircle`, default boundary from compengine is set so small that basically every series I've tried returns 0.
> 2. When checking whether an input is z-scored, the range is so tight that I couldn't find one single series that's z-scored.
> 3. Sample Entropy is calculated very slowly for large time series using loops.  I've tried to improve but didn't make much a difference since I'm not really familiar with the concept about entropy.

I'm not sure how to deal with 1 and 2. I will need to read the definitions in the original paper. sample entropy might need vectorizing to avoid loops as much as possible. Alternatively, use Rcpp.

**Comment by FinYang** [comment]
> > I'm not quite sure how to handle missing values since I don't understand some of the features completely.
> > In `acf()`, na action is na.pass. In any simple arithmetic calculation, such as `mean`, `sd` and `sum`, missing values are removed. In function `cumsum` from function `fluctanal_prop_r1` where it implements fluctuation analysis (I don't know what that means), missing values are treated as 0.
> 
> In `mean`, `sd` and `sum`, use `na.rm=TRUE`. `acf` is problematic as the most obvious option (using complete pairs) is not available. For now, perhaps use `na.action=na.contiguous`.
> 
> Anything using `cumsum`, treating missing as 0s is probably ok for now.

Tried using `na.action=na.contiguous` in `acf`. Apart from errors, functions like "finding the first minimum in autocorrelation " or "finding the first zero corssing in autocorrelation" might need more lags than the length of "Longest Contiguous Stretch of non-NAs", otherwise, I'm not sure if that's different from definition of those functions because it feels `na.contiguous` changes the way it calculates a lot, so I havn't make the change.

**Pull request merged**